### PR TITLE
Pull Request for Review Purpose

### PR DESCRIPTION
**Pull request by mufeili**

### CODE REVIEW

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+import random
+import numpy as np
+import dgl.function as fn
+import dgl
+from dgl.sampling import random_walk, pack_traces
+
+
+# The base class of sampler
+# (TODO): online sampling
+# NOTE: we have to do offline sampling for computing normalization coefficients
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:15]: "coefficients" -> "coefficients at the beginning"

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+class SAINTSampler(Dataset):
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:16]: It's a bit weird to inherit `Dataset` class here as:
1) The dataset should be the output of `load_data` in `train_sampling.py`
2) `Dataset` is just an abstract class with functions like `__getitem__` and `__len__` to be implemented.

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    def __init__(self, dn, g, train_nid, node_budget, num_workers, train, num_subg_train=0, num_subg_norm=10000,
+                 batch_size_norm=200, online=False, num_repeat=50):
+        """
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:19]: Reorder the descriptions of the arguments to have an one-on-one correspondence.

File: .gitignore [hunk 4]
```diff
@@ -161,3 +161,11 @@ cscope.*
...
 .ycm_extra_conf.py
 **.png
+
+# jiahanli
+examples/pytorch/graphsaint/data/*
+examples/pytorch/graphsaint/log/*
+examples/pytorch/graphsaint/subgraphs/*
+examples/pytorch/graphsaint/wandb/*
+*.out
+*.prof
```
Comment (mufeili) [inline .gitignore:171]: Does it work to have a separate `.gitignore` file under `examples/pytorch/graphsaint`?

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        self.edge_counter = th.zeros((self.train_g.num_edges(),))
+        self.prob = None
+        self.num_subg_train = num_subg_train
+        self.num_subg_norm = num_subg_norm
+        self.batch_size_norm = batch_size_norm
+        self.num_workers = num_workers
+        self.train = train
+        self.online = online
+        self.cnt = 0 # count the times sampled subgraphs have been fetched.
+        test = True # TODO: TEST, this flag is set so that we can test sampling time every experiment
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:69]: I guess you can now remove this flag and modify the code accordingly.

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+            aggr_norm, loss_norm = np.load(norm_fn, allow_pickle=True)
+        else:
+            os.makedirs('./subgraphs/', exist_ok=True)
+
+            self.subgraphs = []
+            self.N, sampled_nodes = 0, 0
+            # N: the number of subgraphs sampled in __init__
+
+            # Employ parallelism to speed up the sampling procedure
+            loader = DataLoader(self, batch_size=self.batch_size_norm, shuffle=True, num_workers=self.num_workers,
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:87]: I think `batch_size` here is functionally equivalent to `node_budget` used [here](https://github.com/ljh1064126026/dgl/blob/jiahanli/examples/pytorch/graphsaint/sampler.py#L144) in Tang's code. Why did you use a different number? Is this from the authors' code?

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        :param num_subg_norm: the number of subgraphs sampled in advance for computing normalization coefficients.
+                                Please make sure that num_subg_norm is greater than batch_size_norm so that we can
+                                sample enough subgraphs.
+        :param dn: name of dataset
+        :param g: full graph
+        :param train_nid: ids of training nodes
+        :param node_budget: expected number of sampled nodes. Actually node_budget is the times of sampling nodes
+                            which doesn't mean we sample 6000 different nodes.
+        :param num_workers: number of processes to sample subgraphs using torch.dataloader
+        :param train: True if sampling subgraphs in training procedure
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:33]: I imagine this should also be controlled by `online`?

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+        else:
+            os.makedirs('./subgraphs/', exist_ok=True)
+
+            self.subgraphs = []
+            self.N, sampled_nodes = 0, 0
+            # N: the number of subgraphs sampled in __init__
+
+            # Employ parallelism to speed up the sampling procedure
+            loader = DataLoader(self, batch_size=self.batch_size_norm, shuffle=True, num_workers=self.num_workers,
+                                collate_fn=self.__collate_fn__, drop_last=False)
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:88]: It's kinda weird to pass self to `DataLoader`.

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        if self.train:
+            if self.online is True:
+                subgraph = self.__sample__()
+                return dgl.node_subgraph(self.train_g, subgraph)
+            else:
+                if self.cnt < len(self.subgraphs):
+                    self.cnt += 1
+                    return dgl.node_subgraph(self.train_g, self.subgraphs[self.cnt-1])
+                elif self.online == 'tot':
+                    random.shuffle(self.subgraphs)
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:152]: Is it right to reshuffle each time we get a subgraph?

File: examples/pytorch/graphsaint/sampler_jiahanli.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+    def __getitem__(self, idx):
+        # Only when sampling subgraphs in training procedure and need to utilize sampled subgraphs and we still
+        # have sampled subgraphs can we fetch a subgraph from sampled subgraphs
+        if self.train:
+            if self.online is True:
+                subgraph = self.__sample__()
+                return dgl.node_subgraph(self.train_g, subgraph)
+            else:
+                if self.cnt < len(self.subgraphs):
```
Comment (mufeili) [inline examples/pytorch/graphsaint/sampler_jiahanli.py:148]: When do we reset `self.cnt` to zero if it is fully offline?

File: .gitignore (hunk 4 above)
Comment (ljh1064126026) [reply .gitignore:171]: Nope. This .gitignore file is just for testing some experiments. I'll delete unnecessary files later.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 7 above)
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:33]: Sorry 'bout some confusing comments here. I'll reorganize all comments so that u can clearly see the relationship between this args.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 8 above)
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:88]: Actually I implement the sampler in accordance with the version of @lt610. The sampling methods are different in pre-sampling procedure and training procedure respectively. But we implement the sampler which can handle both of these conditions. So here `DataLoader` utilizes the sampler itself to handle the pre-sampling procedure while the sampler finishes training sampling procedure in `train_sampling.py`.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:87]: Yup. But I'm not sure this factor would affect the performance. Firstly if we'd like to make use of the parallelism of `DataLoader`, we'd better assign `batch_size` larger than 1 so that we can completely utilize each process's ability to sample enough subgraphs. Secondly, u can see that I assign `batch_size_norm` as 200 by default. Because this is how the original author of GraphSAINT handle this param, and I keep the accordance with the author.

[the remark this answers is not in the document]
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:87]: Moreover. Because of the problem u've mentioned, I make the `batch_size_norm` as a hyper parameter which can be adjusted according to the specific circumstance.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 9 above)
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:152]: Actually it does not reshuffle each time. `self.train == True` means the sampling is happening in training procedure. And `self.online == 'tot'` means we employ the completely offline sampling in training procedure. But please note that, when `self.online != True` which means that we employ some other sampling methods instead of online sampling in training procedure, we firstly judge `if self.cnt < len(self.subgraphs)` whatever the `self.online` is. So actually we extract a subgraph from sampled subgraphs when there're some subgraphs left. Only when we extract all sampled subgraphs do we reshuffle.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 10 above)
Comment (ljh1064126026) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:148]: `self.cnt` is to calculate how many subgraphs we've extracted from sampled subgraphs. In `__init__()`, the `self.cnt` is initially set to zero. Then in `__getitem__()`  we add 1 to `self.cnt` when extract a subgraph from sampled subgraphs. If we've already sampled all subgraphs, we shuffle the subgraphs, extract a new one and set `self.cnt` to 1.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:118]: Yes. I've just been spotting the differences between your implementation and Tang's to see which might have caused the performance drop during offline sampling training.

File: examples/pytorch/graphsaint/sampler_jiahanli.py (hunk 9 above)
Comment (mufeili) [reply examples/pytorch/graphsaint/sampler_jiahanli.py:175]: Are you suggesting that you also change self.online during training?

File: examples/pytorch/graphsaint/config.py [hunk 11]
```diff
@@ -5,7 +5,7 @@
         'aggr': 'concat', 'arch': '1-0-1-0', 'dataset': 'ppi', 'dropout': 0, 'edge_budget': 4000, 'gpu': 0, 'length': 2,
         'log_dir': 'none', 'lr': 0.01, 'n_epochs': 1000, 'n_hidden': 512, 'no_batch_norm': False, 'node_budget': 6000,
         'num_subg': 50, 'num_roots': 3000, 'sampler': 'node', 'use_val': False, 'val_every': 1, 'num_workers_sampler': 0,
-        'online': 'tot', 'num_subg_train': 0, 'num_subg_norm': 10000, 'batch_size_norm': 200, 'num_workers': 0
+        'online': 'tot', 'num_subg_train': 0, 'num_subg_norm': 144, 'batch_size_norm': 144, 'num_workers': 0
```
Comment (ljh1064126026) [inline examples/pytorch/graphsaint/config.py:8]: I found the bug that makes offline sampling and author's version cannot work. When I ran the codes of @lt610 I found that the number of offline sampled subgraphs is only 144 while the subgraphs I've sampled to calculated normalization coefficients are too many, which is 200. So I change these params. Previously the fully offline sampling and offline+online sampling (author's version) cannot work. After fixing this bug, they all work.

Comment (ljh1064126026) [reply examples/pytorch/graphsaint/config.py:8]: However, which confuses me is the number of sampled subgraphs in author's codes is 200 as my previous setup. 

The experiment is done on ppi_node_sampler
